src/w3d/renderer/linegroup.cpp

```diff
@@ -0,0 +1,358 @@
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#include "linegroup.h"
+
+#include "camera.h"
+#include "dx8wrapper.h"
+#include "matrix4.h"
+#include "rinfo.h"
+#include "sortingrenderer.h"
+
+LineGroupClass::LineGroupClass() :
+    m_startLoc(nullptr),
+    m_endLoc(nullptr),
+    m_diffuseBuffer(nullptr),
+    m_tailDiffuseBuffer(nullptr),
+    m_altBuffer(nullptr),
+    m_sizeBuffer(nullptr),
+    m_UCoordsBuffer(nullptr),
+    m_lineCount(0),
+    m_texture(nullptr),
+    m_flags(0),
+    m_defaultSize(0),
+    m_defaultColor(1, 1, 1),
+    m_defaultAlpha(1),
+    m_defaultUCoord(0),
+    m_defaultTailDiffuse(0, 0, 0, 0),
+    m_mode(TETRAHEDRON)
+{
+}
+
+LineGroupClass::~LineGroupClass()
+{
+    Ref_Ptr_Release(m_startLoc);
+    Ref_Ptr_Release(m_endLoc);
+    Ref_Ptr_Release(m_diffuseBuffer);
+    Ref_Ptr_Release(m_tailDiffuseBuffer);
+    Ref_Ptr_Release(m_altBuffer);
+    Ref_Ptr_Release(m_sizeBuffer);
+    Ref_Ptr_Release(m_UCoordsBuffer);
+    Ref_Ptr_Release(m_texture);
+}
+
+void LineGroupClass::Set_Arrays(ShareBufferClass<Vector3> *start_locs,
+    ShareBufferClass<Vector3> *end_locs,
+    ShareBufferClass<Vector4> *diffuse,
+    ShareBufferClass<Vector4> *tail_diffuse,
+    ShareBufferClass<unsigned int> *alt,
+    ShareBufferClass<float> *sizes,
+    ShareBufferClass<float> *u_coords,
+    int count)
+{
+    captainslog_assert(start_locs);
+    captainslog_assert(end_locs);
+    captainslog_assert(start_locs->Get_Count() == end_locs->Get_Count());
+    captainslog_assert(!diffuse || start_locs->Get_Count() == diffuse->Get_Count());
+    captainslog_assert(!alt || start_locs->Get_Count() == alt->Get_Count());
+    captainslog_assert(!sizes || start_locs->Get_Count() == sizes->Get_Count());
+    captainslog_assert(!u_coords || start_locs->Get_Count() == u_coords->Get_Count());
+    captainslog_assert(!tail_diffuse || start_locs->Get_Count() == tail_diffuse->Get_Count());
+
+    Ref_Ptr_Set(m_startLoc, start_locs);
+    Ref_Ptr_Set(m_endLoc, end_locs);
+    Ref_Ptr_Set(m_diffuseBuffer, diffuse);
+    Ref_Ptr_Set(m_tailDiffuseBuffer, tail_diffuse);
+    Ref_Ptr_Set(m_altBuffer, alt);
+    Ref_Ptr_Set(m_sizeBuffer, sizes);
+    Ref_Ptr_Set(m_UCoordsBuffer, u_coords);
+
+    if (m_altBuffer != nullptr) {
+        m_lineCount = count;
+    } else {
+
+        if (m_startLoc != nullptr && count < 0) {
+            count = m_startLoc->Get_Count();
+        }
+
+        m_lineCount = count;
+    }
+}
+
+void LineGroupClass::Set_Texture(TextureClass *texture)
+{
+    if (texture != nullptr) {
+        texture->Add_Ref();
+    }
+    m_texture = texture;
+}
+
+void LineGroupClass::Render(RenderInfoClass &rinfo)
+{
+#ifdef BUILD_WITH_D3D8
+    if (m_lineCount > 0) {
+        m_shader.Set_Cull_Mode(ShaderClass::CULL_MODE_DISABLE);
```

Comment:
Think you've made a mistake here `CULL_MODE_ENABLE`

Reply:
indeed..